### PR TITLE
Silence traces when logging langchain models

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/12210?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12210/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12210
```

</p>
</details>

### What changes are proposed in this pull request?

This is an extension of https://github.com/mlflow/mlflow/pull/12167, where we disable trace during signature inference for logging models. While it worked well for normal pyfunc models, we observed a few traces were still generated when saving some particular model types such as LangChain model-as-code models and Chat Models. These models require a few additional prediction calls during logging, not only the one for signature inference.

To ensure the coverage of the suppression, this PR implements `@trace_disabled` decorator instead of context manager, and apply it to the entire `log_model()` and `save_model()` function.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests 

### Before
Screenshot 2024-06-03 at 13 58 00

### After 

Screenshot 2024-06-03 at 13 58 53

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/types/utils.py [hunk 1]
```diff
@@ -238,30 +238,6 @@ def _infer_schema(data: Any) -> Schema:
     """
     from scipy.sparse import csc_matrix, csr_matrix
 
-    # List[Dict[str, Union[DataType, List, Dict]]]
```
Comment (B-Step62) [inline mlflow/types/utils.py:241]: note: Unrelated to this PR, but we found this warning is no-longer required, during the same discussion with PM.

File: mlflow/types/utils.py [hunk 2]
```diff
@@ -238,30 +238,6 @@ def _infer_schema(data: Any) -> Schema:
...
-    # 1  some sentence  [id1, id2]  {'key': 'value', 'key2': 'value2'}
-    # inferred schema -->
-    # Schema([ColSpec(type=DataType.string, name='output'),
-    #        ColSpec(type=Array(dtype=DataType.string), name='ids'),
-    #        ColSpec(type=Object([Property(name='key', dtype=DataType.string),
-    #                             Property(name='key2', dtype=DataType.string, required=False)]
-    #               ), name='dict')])
-    if isinstance(data, dict) and any(not isinstance(value, str) for value in data.values()):
-        # TODO: Add a link to docs/examples
-        _logger.info(
```
Comment (B-Step62) [inline mlflow/types/utils.py:258]: note: Unrelated to this PR, but we found this message is no-longer required, during the same discussion with PM.

File: mlflow/tracing/provider.py [hunk 3]
```diff
@@ -136,20 +136,39 @@ def enable():
...
+            @trace_disabled
+            def f():
+                with mlflow.start_span("my_span") as span:
+                    span.set_attribute("my_key", "my_value")
+
+                return
+
+
+            # This function will not generate any trace
+            f()
```
Comment (harupy) [inline mlflow/tracing/provider.py:154]:
```suggestion
        @trace_disabled
        def f():
            with mlflow.start_span("my_span") as span:
                span.set_attribute("my_key", "my_value")

            return


        # This function will not generate any trace
        f()
```

File: mlflow/tracking/client.py [hunk 4]
```diff
@@ -578,8 +578,12 @@ def start_trace(
                 name, experiment_id=experiment_id
             )
             request_id = get_otel_attribute(otel_span, SpanAttributeKey.REQUEST_ID)
-
             mlflow_span = create_mlflow_span(otel_span, request_id, span_type)
+
+            # # If the span is a no-op span i.e. tracing is disabled, do nothing
+            if isinstance(mlflow_span, NoOpSpan):
```
Comment (B-Step62) [inline mlflow/tracking/client.py:584]: note: This is for eliminating the following warning
> 2024/06/03 04:57:21 WARNING mlflow.tracking.client: Failed to start trace RunnableSequence: INVALID_PARAMETER_VALUE: Invalid request id: None in GetTraceInfo request. Request id must map to a trace.. For full traceback, set logging level to debug.

File: mlflow/types/utils.py (hunk 2 above)
Comment (harupy) [reply mlflow/types/utils.py:258]: Can we isolate this in a separate PR? We don't want to revert this when we need to revert this PR.

Comment (B-Step62) [reply mlflow/types/utils.py:258]: Sure, will file a new PR for it

Comment (B-Step62) [reply mlflow/types/utils.py]: Filed https://github.com/mlflow/mlflow/pull/12212

**Review by serena-ruan — APPROVED** [review]
LGTM!

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**